### PR TITLE
#4819 Marketplace Publish A3: Add Edit Pending Publish Modal

### LINKED ISSUE #4819: Marketplace Publish A3: Add Edit Pending Publish Modal
See Slice 3 in this spec with mockups:

<https://www.notion.so/pixiebrix/Public-Marketplace-Publishing-for-Designated-Creators-f3adaf2f0f8f4cc5929643abe9bb3348>

<p></p>

- While a blueprint is in the state of pending publish, replace the “Publish to Marketplace” with the “Edit Pending Publish” modal\.

- If the user clicks “Cancel Publish”, close the “Edit Pending Publish” modal and display the confirmation\.

- If the pending publish is canceled, the public link is removed\.

<p></p>

Week 50 \- 5 story points

Week 51 \- 1 story point remaining

<p></p>

<p></p>

### PR DESCRIPTION
**Pull request by BALEHOK**
## What does this PR do?

- Closes #4819
- Refactors the Blueprint Modals to always stay in the React tree.

## Reviewer Tips

The approach with the modals is:
1. a ModalLayout receives a `show` flag via props.
2. ModalLayout passes the flag to the bootstrap `Modal`.
3. ModalLayout conditionally renders the modal content.

## Discussion

These are the screenshots of all the Blueprint modals. They have certain differences (displaying Blueprint name, order of destructive/non-destractive buttons). Maybe we can make them more uniform?

Share with teams:
Screenshot 2022-12-16 at 12 03 54 PM

No permission to share:
Screenshot 2022-12-16 at 12 04 26 PM

Publish to marketplace ([spec](https://www.notion.so/pixiebrix/Public-Marketplace-Publishing-f3adaf2f0f8f4cc5929643abe9bb3348#e80f2d2fb6414f1d96ee459023215579)):
Screenshot 2022-12-16 at 12 03 41 PM

No permission to publish:
Screenshot 2022-12-16 at 12 04 13 PM

Edit pending publish:
Screenshot 2022-12-16 at 4 13 04 PM

Confirm cancel publish:
Screenshot 2022-12-20 at 4 44 48 pm

## Demo

https://www.loom.com/share/0c8f0d72477d43b8b385ddb8eabd1d2d

## Checklist

- [x] Add tests
- [ ] Run Storybook and manually confirm that all stories are working
- [x] Designate a primary reviewer @twschiller

### CODE REVIEW

**Comment by twschiller** [comment]
@mnholtz could you help review this PR?

@corinnemayans @icugideon did you have any feedback on @BALEHOK's discussion comment on the modal layout? For the "Share with Teams" modal, I just had copied the layout of the Google Docs sharing modal

**Comment by corinnemayans** [comment]
There are some design pattern things that we ought to clean up, but we'll wait and do that more holistically. (These are things like button order/color/size, heading sizes and use, etc.)

The one quick fix would be the blueprint name font size in Publish to Marketplace. It looks like it is using a header size that is larger than the modal title. It ought to be no larger than the modal title. We'll better define the various header sizes and their use in future patterns, but for now, a smaller header or no header would work. This isn't a blocker to closing this issue, though.

**Comment by BALEHOK** [comment]
> for now, a smaller header or no header would work

I'd rather remove the header from the 2 modals to make them more consistent.

File: src/options/pages/blueprints/modals/shareModals/PublishRecipeContent.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      notify.success("Shared brick");
+      closeModal();
+      refetchRecipes();
+    } catch (error) {
+      if (
+        isSingleObjectBadRequestError(error) &&
+        error.response.data.config?.length > 0
+      ) {
+        setError(error.response.data.config.join(" "));
+      } else {
```
Comment (BLoe) [inline src/options/pages/blueprints/modals/shareModals/PublishRecipeContent.tsx:77]: I don't remember seeing this error logic anywhere else. Do you know if this is something we could or should move inside the `getErrorMessage()` logic itself?

File: src/options/pages/blueprints/modals/shareModals/PublishRecipeContent.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        <p>
+          As soon as you Submit, the public link below will work for anyone, so
+          you can start sharing right away!
+        </p>
+
+        <p className="mb-1">Public link to share:</p>
+        <ActivationLink blueprintId={blueprintId} />
+      </Modal.Body>
+      <Modal.Footer>
+        <Button variant="link" disabled={disableButtons} onClick={closeModal}>
```
Comment (BLoe) [inline src/options/pages/blueprints/modals/shareModals/PublishRecipeContent.tsx:119]: Does the cancel button really need to wait for packages to be fetched before it's active?

File: src/options/pages/blueprints/modals/shareModals/PublishRecipeModals.test.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    }),
+  });
+});
+
+afterEach(() => {
+  jest.clearAllMocks();
+});
+
+function mockHooks() {
+  (useRecipe as jest.Mock).mockReturnValue({
```
Comment (BLoe) [inline src/options/pages/blueprints/modals/shareModals/PublishRecipeModals.test.tsx:66]: Can we use `mockImplementation` here to have it pull the current values each time instead of re-mocking the hook each time? Then, we wouldn't have to call this in every test.

File: src/options/pages/blueprints/modals/shareModals/PublishRecipeModals.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  if (!recipe.sharing.public) {
+    return <PublishRecipeContent />;
+  }
+
+  if (isRecipePendingPublish(recipe, listings)) {
+    return cancelingPublish ? <CancelPublishContent /> : <EditPublishContent />;
+  }
+
+  // Normally we shouldn't get here, because the Publish action is not displayed for a published blueprint
```
Comment (BLoe) [inline src/options/pages/blueprints/modals/shareModals/PublishRecipeModals.tsx:55]: Maybe it could happen if the user has two instances of the blueprints page open at the same time and shares a blueprint on one page and then tries to do the same thing on the other page?

File: src/options/pages/blueprints/modals/shareModals/ShareRecipeModalBody.tsx [hunk 5]
```diff
@@ -69,7 +68,7 @@ const AddATeamMenuList = createMenuListWithAddButton(
   "https://app.pixiebrix.com/teams/create"
 );
 
-const ShareRecipeModal: React.FunctionComponent = () => {
+const ShareRecipeModalBody: React.FunctionComponent = () => {
```
Comment (BLoe) [inline src/options/pages/blueprints/modals/shareModals/ShareRecipeModalBody.tsx:71]: Ok, we're using `ModalBody` in the name here for this one, so let's definitely use the same naming convention for the other modals.

File: src/options/pages/blueprints/useInstallableViewItems.tsx [hunk 6]
```diff
@@ -118,33 +118,43 @@ function useInstallableViewItems(installables: Installable[]): {
...
+              installable,
+              organizations,
+              scope,
+              installedExtensions,
+            }),
+            isPublished:
+              listingsQuery.data == null
+                ? false
+                : // eslint-disable-next-line security/detect-object-injection -- packageId is a registry id
+                  Boolean(listingsQuery.data[packageId]),
```
Comment (BLoe) [inline src/options/pages/blueprints/useInstallableViewItems.tsx:139]: Can we just use `?.[]` safe access here instead of the full ternary?
```
Boolean(listingsQuery.data?.[packageId])
```

**Review by BLoe — APPROVED** [review]
Looks fine to me, just had some questions on naming and some nit picks

File: src/options/pages/blueprints/modals/shareModals/PublishContentLayout.tsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+const editorRoles = new Set<number>([UserRole.admin, UserRole.developer]);
+
+const sortOrganizations = (organizations: Organization[]) =>
+  sortBy(organizations, (organization) => organization.name);
+
+type PublishModalLayoutProps = React.PropsWithChildren<{
+  title: string;
+}>;
+
+const PublishContentLayout: React.FunctionComponent<
```
[the remark this answers is not in the document]
Comment (BALEHOK) [reply src/options/pages/blueprints/modals/shareModals/PublishContentLayout.tsx:46]: This situation is different. The naming of the modals doesn't follow the naming pattern of components, but reflects what one or another component is responsible for. Let me explain the modals of this branch a bit.
We want to show animation, but don't want to render the content of a hidden modal. This is accomplished via a component with always shown `Modal` and conditionally rendered content.
This component represents a sort of template or a layout for a modal. See [ModalLayout.tsx](https://github.com/pixiebrix/pixiebrix-extension/pull/4835/files#diff-ba6427d791e24ff59dd0d073664183bb10d7d7abf9afeb4f2485b1ce37af3793) and its usage, for example [ExtensionLogsModal.tsx](https://github.com/pixiebrix/pixiebrix-extension/pull/4835/files#diff-226ca033c29a0ab5921ac9171ac1b3f32ba1150f4c74bc7af8c13755764a2403R47).
`ModalLayout` render the `Header` and children. In most cases children is the body of the modal, hence the naming for `ShareRecipeModalBody` used in [ShareRecipeModal.tsx](https://github.com/pixiebrix/pixiebrix-extension/pull/4835/files#diff-86a476af166736319db1ba54ea1bea376d79642b6d5eb3341c091a721a27dc24R41).
Publish modals require a bit more flexibility. [PublishRecipeModals](https://github.com/pixiebrix/pixiebrix-extension/pull/4835/files#diff-f182aea32df5d7fbcc1e91ce8a88217fb049c087449e2f6745df9f7697d64c86R59) is the root. It will decide what content to render in the modal, not just a body. So everything (apart from a `Loader`) returned from the [ModalContentSwitch](https://github.com/pixiebrix/pixiebrix-extension/pull/4835/files#diff-f182aea32df5d7fbcc1e91ce8a88217fb049c087449e2f6745df9f7697d64c86R32) has a `Content` postfix. Then all those content components have a certain layout. The logic here is similar to `ModalLayout.tsx` but it's about internals of a modal, how the content (Header, Body, Footer) will be rendered. This is why the component that knows how to layout a content of a publish modal is called [PublishContentLayout](https://github.com/pixiebrix/pixiebrix-extension/pull/4835/files#diff-cc53bde6813e9b186e2bff0e1e6a7564c12fc90961c9451993a368554a39af67).

To sum up, there are components which are layout templates - `*Layout` postfix. There're components which render modals' body - `*Body` postfix. There're components which render modals' content (Header, Body, Footer - everything included in a modal) - `*Content` postfix.

I'm happy to discuss the renaming further. Just want to note, that the issue has been open for a week and I'll be OOO during the next week.

File: src/options/pages/blueprints/modals/shareModals/PublishContentLayout.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+import { type Organization, UserRole } from "@/types/contract";
+import { useRecipe } from "@/recipes/recipesHooks";
+import { RequireScope } from "@/auth/RequireScope";
+
+const editorRoles = new Set<number>([UserRole.admin, UserRole.developer]);
+
+const sortOrganizations = (organizations: Organization[]) =>
+  sortBy(organizations, (organization) => organization.name);
+
+type PublishModalLayoutProps = React.PropsWithChildren<{
```
[the remark this answers is not in the document]
Comment (BALEHOK) [reply src/options/pages/blueprints/modals/shareModals/PublishContentLayout.tsx:42]: absolutely, thanks!

File: src/options/pages/blueprints/modals/shareModals/PublishRecipeContent.tsx (hunk 1 above)
Comment (BALEHOK) [reply src/options/pages/blueprints/modals/shareModals/PublishRecipeContent.tsx:77]: This is copied from the (former) `ShareRecipeModal` with slight modification - [here](https://github.com/pixiebrix/pixiebrix-extension/pull/4835/files#diff-04da910336bd855157168db99ee6c70394325f40daab357289c4cf0d37bbaa89) at line 121.

If you think it is worth moving this logic to the `getErrorMessage`, let's handle it separately, I think this PR has enough changes. Could you create a ticket?

File: src/options/pages/blueprints/modals/shareModals/PublishRecipeContent.tsx (hunk 2 above)
Comment (BALEHOK) [reply src/options/pages/blueprints/modals/shareModals/PublishRecipeContent.tsx:119]: Nope. Good catch.

File: src/options/pages/blueprints/modals/shareModals/PublishRecipeModals.tsx (hunk 4 above)
Comment (BALEHOK) [reply src/options/pages/blueprints/modals/shareModals/PublishRecipeModals.tsx:55]: Publishing is not instant, it requires approval from admins. This situation seems to be unlikely. Anyway, the handler is here :)

File: src/options/pages/blueprints/useInstallableViewItems.tsx (hunk 6 above)
Comment (BALEHOK) [reply src/options/pages/blueprints/useInstallableViewItems.tsx:139]: changed in the follow up PR: https://github.com/pixiebrix/pixiebrix-extension/pull/4861/files#diff-b45edee50926f99414054c52e1c0f24677c85f5b6bb2ae456bd03f295efcbf94R136

File: src/options/pages/blueprints/modals/shareModals/ShareRecipeModalBody.tsx (hunk 5 above)
Comment (BALEHOK) [reply src/options/pages/blueprints/modals/shareModals/ShareRecipeModalBody.tsx:71]: see the discussion above

**Comment by BALEHOK** [comment]
Not merging this, because the naming discussion is not resolved.

File: src/options/pages/blueprints/modals/shareModals/PublishContentLayout.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (BLoe) [reply src/options/pages/blueprints/modals/shareModals/PublishContentLayout.tsx:46]: I guess it's fine if you have a system for how everything is named, I don't want to hold up the PR over this.

**Comment by BLoe** [comment]
I'm fine with merging this, I don't think that naming discussion is worth blocking the PR over. Thanks for following up on it!

**Comment by BLoe** [comment]
A few jsDoc comments on the various Layout/Content components might go a long way to helping understand the structure of everything though.
